server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -835,6 +892,9 @@ public static ClusterState readFrom(StreamInput in, DiscoveryNode localNode) thr
         builder.metadata = Metadata.readFrom(in);
         builder.routingTable = RoutingTable.readFrom(in);
         builder.nodes = DiscoveryNodes.readFrom(in, localNode);
+        if (in.getTransportVersion().onOrAfter(TransportVersion.V_8_8_0)) {
+            builder.transportVersions(in.readMap(StreamInput::readString, TransportVersion::readVersion));
+        }
```

Comment:
I think there's a subtle BwC problem here.

This logic means that a cluster state received from a pre-8.8.0 master will have an empty transport-versions map. When the first post-8.8.0 master is elected, it will therefore start with this empty map, and it won't populate it because `NodeJoinExecutor` treats join requests from known nodes as no-ops.

I think we need to lean on the existing node-version barrier to address this: while there are any pre-8.8.0 nodes in the cluster we should keep this map empty, meaning that the transport-version barrier is inactive; once the last pre-8.8.0 node leaves and the node-version barrier advances we can do an explicit step to populate this map completely, putting the transport-version barrier in place.

Reply:
Where do we get the complete transportversions info from when we move from pre-8.8.0 to 8.8.0?

Reply:
Could we add it to `TransportInfo` so we can use the nodes info action? Otherwise I guess we'll need a new action.

Reply:
This is getting to be a much more complicated change, is it worth reconsidering https://github.com/elastic/elasticsearch/pull/93818 - storing the transportversion in the discoverynode, but only accessing it through ClusterState. That way we can propagate all this info in the cluster, without having to add all the ceremony around managing this extra piece of state.

Reply:
I think that change has the same BwC issue tho, and in fact it's better hidden which makes it worse. It's not really important where this info lives in the cluster state, the rolling-upgrade problem is more fundamental than that.

Reply:
But in storing it in DiscoveryNode we can provide a default based on the Version (as, pre 8.8.0, there's a one-to-one mapping between the two). So we can still calculate it from previous nodes, because we have a `Version` to base it off.

Reply:
Ah ok. But then we can do this here too right?

I really don't want `DiscoveryNode` to have a `TransportVersion`. I expect it will lead to bugs.

Reply:
What type of bugs do you anticipate? What about DiscoveryNode makes it not appropriate for information about each node? In particular, it already has version, why is transport version not appropriate?

Reply:
`TransportVersion` is a property of a direct transport connection between two nodes, but if we make it a property of the `DiscoveryNode` then it seems more likely that at some point we'll accidentally write some logic that only works for nodes to which we are directly connected, which is _almost_ always the case but breaks in rare situations such as cross-cluster setups involving proxies during a rolling ugprade.

`DiscoveryNode` is already a weird mash-up of random stuff related to a node. I'd like to make it smaller, not larger.

I'm not going to block progress on this front tho. I don't think it's any easier to attach it to `DiscoveryNode` than it is to keep it separate, and I see conceptual advantages to keeping it separate, but maybe I'm missing something.

Reply:
I've pushed a change to infer the transport version from the node version where possible